### ISSUE TITLE
Any plans to implement recurrent layers?

### ISSUE THREAD

**Issue by LarocheC** [BobMcDear/attorch#11]
Hi,
First, thanks for the great project!
I wondered if there are any plans to implement recurrent layers (such as RNN, LSTM, or GRU) using Triton. I don’t have experience with Triton yet, I’d be interested in contributing if there’s a way I can help.

Thanks!

**Comment by BobMcDear** [comment]
Hello,

Thank you for your interest in attorch! If the sequence is operated over using a plain Python for loop, adding recurrent cells shouldn't be particularly challenging since they're little but linear layers and non-linearities. However, the overhead of this approach can be substantial, especially for smaller networks. Triton offers associative scans, which can be leveraged to efficiently implement linear RNNs, but gated ones like LSTMs and GRUs would be trickier to develop. It's not impossible, but it'll take a bit of work to achieve PyTorch-like performance. I'm busy at the moment, but I'll see if I can start chipping away at it. You're more than welcome to submit a pull request should you decide to learn Triton.

**Comment by LarocheC** [comment]
Hi,

I totally agree. The [GRU cell](https://pytorch.org/docs/stable/generated/torch.nn.GRUCell.html) and [LSTM cell](https://pytorch.org/docs/stable/generated/torch.nn.LSTMCell.html) would be fairly easy to implement but slow when operating over a sequence. 

 I have seen someone trying to (re)do a CUDA kernel for the GRU here: https://github.com/kevinghst/GRU-CUDA, but it does not look very good when processing sequence to get close to PyTorch performance. 


I was wondering what type of optimization has to be done to speed up the recurrence over the sequence axis? 

A naïve idea was to pre-compute the "input to hidden" for all timesteps, then loop over the timesteps for the rest of the computation. (example code in Pytorch) 

```
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np

class CustomGRU(nn.Module):
    def __init__(self, input_size, hidden_size, batch_first=True):
        super(CustomGRU, self).__init__()
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.batch_first = batch_first
        
        self.weight_ih = nn.Parameter(torch.Tensor(3 * hidden_size, input_size))
        self.weight_hh = nn.Parameter(torch.Tensor(3 * hidden_size, hidden_size))
        self.bias_ih = nn.Parameter(torch.Tensor(3 * hidden_size))
        self.bias_hh = nn.Parameter(torch.Tensor(3 * hidden_size))
        
        self.init_weights()
    
    def init_weights(self):
        for param in self.parameters():
            nn.init.uniform_(param, -0.1, 0.1)
    
    def forward(self, x, h_0=None):
        # x shape: (batch_size, seq_len, input_size) if batch_first else (seq_len, batch_size, input_size)
        if self.batch_first:
            batch_size, seq_len, _ = x.size()
        else:
            seq_len, batch_size, _ = x.size()
            x = x.permute(1, 0, 2)  # Convert to batch_first for easier handling
        
        if h_0 is None:
            h_0 = torch.zeros(batch_size, self.hidden_size, device=x.device)
        

        # Precompute input for all timesteps
        x_transformed = F.linear(x, self.weight_ih, self.bias_ih)  # Shape: [batch, seq_len, 3*hidden_size]

        h_t = h_0
        hidden_seq = []

        for t in range(seq_len):
            # Get precomputed values for current timestep
            x_t = x_transformed[:, t, :]

            # Compute reset & update gates in one fused operation
            r_z_t = torch.sigmoid(F.linear(h_t, self.weight_hh[:2*self.hidden_size], self.bias_hh[:2*self.hidden_size]) + x_t[..., :2*self.hidden_size])
            r_t, z_t = r_z_t.chunk(2, dim=-1)  # Efficiently split after activation

            # Compute new memory content in another fused operation
            n_t = torch.tanh(F.linear(h_t, self.weight_hh[2*self.hidden_size:], self.bias_hh[2*self.hidden_size:]) * r_t + x_t[..., 2*self.hidden_size:])
            # Hidden state update
            h_t = (1 - z_t) * n_t + z_t * h_t

            hidden_seq.append(h_t)
        
        hidden_seq = torch.stack(hidden_seq, dim=1)
        
        if not self.batch_first:
            hidden_seq = hidden_seq.permute(1, 0, 2)
        
        return hidden_seq, h_t

def test_custom_gru():
    # Set random seed for reproducibility
    torch.manual_seed(42)
    np.random.seed(42)
    
    # Parameters
    batch_size = 3
    seq_len = 5
    input_size = 10
    hidden_size = 8
    
    # Create input data
    x = torch.randn(batch_size, seq_len, input_size)
    h_0 = torch.randn(batch_size, hidden_size)
    
    # Initialize custom GRU
    custom_gru = CustomGRU(input_size, hidden_size)
    
    pytorch_gru = nn.GRU(input_size, hidden_size, batch_first=True)
    
    # Map custom GRU parameters to PyTorch GRU parameters
    with torch.no_grad():
        # PyTorch GRU has weight_ih_l0 and weight_hh_l0 matrices
        pytorch_gru.weight_ih_l0.copy_(custom_gru.weight_ih)
        pytorch_gru.weight_hh_l0.copy_(custom_gru.weight_hh)
        pytorch_gru.bias_ih_l0.copy_(custom_gru.bias_ih)
        pytorch_gru.bias_hh_l0.copy_(custom_gru.bias_hh)
    
    # Forward pass custom GRU
    custom_output, custom_h_n = custom_gru(x, h_0)
    
    # Forward pass PyTorch GRU
    pytorch_output, pytorch_h_n = pytorch_gru(x, h_0.unsqueeze(0))
    pytorch_h_n = pytorch_h_n.squeeze(0)
    
    # Compare outputs
    output_diff = torch.abs(custom_output - pytorch_output).max().item()
    h_n_diff = torch.abs(custom_h_n - pytorch_h_n).max().item()
    
    print(f"Maximum difference in outputs: {output_diff:.6f}")
    print(f"Maximum difference in final hidden state: {h_n_diff:.6f}")
    
    # Check if custom implementation matches PyTorch's
    output_match = torch.allclose(custom_output, pytorch_output, rtol=1e-4, atol=1e-4)
    h_n_match = torch.allclose(custom_h_n, pytorch_h_n, rtol=1e-4, atol=1e-4)
    
    print(f"Output tensors match: {output_match}")
    print(f"Final hidden states match: {h_n_match}")
    
    return output_match and h_n_match

if __name__ == "__main__":
    test_result = test_custom_gru()
    print(f"Test {'passed' if test_result else 'failed'}!")
```

**Comment by BobMcDear** [comment]
Triton and CUDA operate on different layers of abstraction, and maximizing the performance of the latter is generally much more difficult. For example, writing a CUDA kernel for efficient matrix multiplication can be [incredibly complicated](https://github.com/leimao/CUDA-GEMM-Optimization), whereas Triton achieves performance on par with CUBLAS [in a few dozen lines of code](https://triton-lang.org/main/getting-started/tutorials/03-matrix-multiplication.html). As such, I expect RNNs to be a great deal simpler to implement in Triton, but I can't predict where the bottlenecks would be and what kinds of optimizations will be necessary without running some experiments first.

**Comment by LarocheC** [comment]
Thanks for the comments. After some research I found this:

https://github.com/tile-ai/tilelang

Maybe you could be interested in having a look at it.

**Comment by BobMcDear** [comment]
Thank you for sharing this, it looks interesting. tile-lang appears to be built on TVM; in my experience, deep learning kernels tend to be more difficult to develop efficiently in TVM than Triton, and the latter already has quite an impressive ecosystem which the former lacks.

**Comment by BobMcDear** [comment]
For now, I'm closing this issue because efficient recurrent kernels will require significant effort to develop in Triton, especially since there's no reference, and I can't commit myself to that at the moment. Please feel free to reopen it though if you had any suggestions or got the chance to implement a rough baseline.
